domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/gtfs/calendardates/CalendarDate.java

```diff
@@ -0,0 +1,156 @@
+            this.serviceId = serviceId;
+            return this;
+        }
+
+        /**
+         * Sets field date value and returns this
+         *
+         * @param date date when service exception occurs
+         * @return builder for future object creation
+         */
+        public CalendarDateBuilder date(@NotNull final LocalDateTime date) {
+            this.date = date;
+            return this;
+        }
+
+        /**
+         * Sets field exceptionType value and returns this
+         *
+         * @param exceptionType indicates whether service is available on the date specified in the date field
+         * @return builder for future object creation
+         */
+        public CalendarDateBuilder exceptionType(@NotNull final Integer exceptionType) {
+            this.exceptionType = ExceptionType.fromInt(exceptionType);
+            this.originalExceptionTypeInteger = exceptionType;
+            return this;
+        }
+
+        /**
+         * Returns an entity representing a row from calendar_dates.txt if the requirements from the official GTFS
+         * specification are met. Otherwise, method returns an entity representing a list of notices.
+         *
+         * @return entity representing a row from calendar_dates.txt if the requirements from the official GTFS
+         * specification are met. Otherwise, method returns an entity representing a list of notices.
+         */
+        public EntityBuildResult<?> build() {
+            noticeCollection.clear();
+            if (serviceId == null || date == null || exceptionType == null) {
+                if (serviceId == null) {
+                    noticeCollection.add(new MissingRequiredValueNotice("calendar_dates.txt",
+                            "service_id", serviceId));
+                }
+                if (date == null) {
+                    noticeCollection.add(new MissingRequiredValueNotice("calendar_dates.txt",
+                            "date", serviceId));
+                }
+                if (exceptionType == null) {
+                    if (originalExceptionTypeInteger == null) {
+                        noticeCollection.add(new MissingRequiredValueNotice("calendar_dates.txt",
+                                "exception_type", serviceId));
+                    } else if (!ExceptionType.isEnumValueValid(originalExceptionTypeInteger)) {
+                        noticeCollection.add(new UnexpectedEnumValueNotice("calendar_dates.txt",
+                                "exception_type", serviceId, originalExceptionTypeInteger));
+                    }
+                }
+                return new EntityBuildResult<>(noticeCollection);
+            } else {
+                return new EntityBuildResult<>(new CalendarDate(serviceId, date, exceptionType));
+            }
+        }
+    }
+
+    public String getCalendarDateMappingKey() {
+        return getServiceId() + getDate().toString() + getExceptionType().toString();
```

Comment:
One note on the composite key here - thinking through this further, I believe we could get away with just having `getServiceId() + getDate().toString()` be the composite key. In both use cases of calendar_dates.txt (outlined [here](https://github.com/MobilityData/gtfs-validator/issues/189#issuecomment-626985584)) you shouldn't have the same service_id and date combination more than once - this would imply that you're adding and removing the same service on the same date.

The Python validator actually has a rule that checks to make sure there isn't more than one service_id + date entry with the same values - https://github.com/MobilityData/gtfs-validator/issues/189#issuecomment-626985584.

@timMillet Could you confirm this?

Reply:
Would considering just these two fields prevent us from detecting that a row from calendar.txt with same service_id, date and exception_type is already defined? Unless we check the 3rd element when trying to add a new service to the GtfsDataRepository.

Reply:
Correct, if we use two fields for the key, you'd need to fetch the existing record with the same key from the HashMap and do a manual comparison with the third field to see if the third field matched.

But, you can't do the reverse - key on three fields and then check if two fields match.

For example, this is wrong in GTFS (two field collision):

~~~
service_id,date,exception_type
ABCD,20200509,1
ABCD,20200509,2
~~~

And so is this:

~~~
service_id,date,exception_type
ABCD,20200509,1
ABCD,20200509,1
~~~

By keying on `getServiceId() + getDate().toString()` you can catch both situations (by manually checking the 3rd element), but when keying on `getServiceId() + getDate().toString() + getExceptionType().toString()` you can only catch the 2nd (because there would be no collision in the HashMap key for the first set of data).

One question is if we want to split these out as two different errors, or leave them under the same error. If they are the same error (e.g., "Invalid calendar_dates.txt data"), then you don't even need to check if the 3rd element matches.

Reply:
Thanks for the explanation, I understand the issue better now. Could we in a first approach consider the errors to be the same, them depending on the rules to implement make them different?

Reply:
IMHO yes. I'd suggest using `getServiceId() + getDate().toString()` as the key, and if there are collisions it's an error. This matches the current implementation in the Python validator and it's likely we won't need to go deeper than that.

Reply:
> you shouldn't have the same service_id and date combination more than once - this would imply that you're adding and removing the same service on the same date
I agree